### ISSUE TITLE
Can't establish SSL connection after changing project name

### ISSUE THREAD

**Issue by amsoedal** [dotnet/runtime#56543]
Originally opened here (https://github.com/kubernetes-client/csharp/issues/673), but this seems to be a dotnet issue.

After making a small change to our project and rebuilding it from our pipeline, I started reproducing the below exception consistently on MacOS. The weird thing is that if I step backwards one commit and build the DLLs from our pipeline again, the issue goes away. The commit that changed the behavior was simply renaming the project, and as far as I can tell nothing else changed -- I've double-checked both the pipeline as well as the commit and I'm at a loss. This never reproduces when I build locally. 

```
System.Net.Http.HttpRequestException: 
{
    "StackTrace": "   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.GetHttp2ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)\n   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at k8s.WatcherDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\n   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)\n   at k8s.Kubernetes.ListNamespacedPodWithHttpMessagesAsync(String namespaceParameter, Nullable`1 allowWatchBookmarks, String continueParameter, String fieldSelector, String labelSelector, Nullable`1 limit, String resourceVersion, String resourceVersionMatch, Nullable`1 timeoutSeconds, Nullable`1 watch, String pretty, Dictionary`2 customHeaders, CancellationToken cancellationToken)\n   at k8s.KubernetesExtensions.ListNamespacedPodAsync(IKubernetes operations, String namespaceParameter, Nullable`1 allowWatchBookmarks, String continueParameter, String fieldSelector, String labelSelector, Nullable`1 limit, String resourceVersion, String resourceVersionMatch, Nullable`1 timeoutSeconds, Nullable`1 watch, String pretty, CancellationToken cancellationToken)\n   at Microsoft.BridgeToKubernetes.Common.Kubernetes.KubernetesClient.<>c__DisplayClass29_0.<<ListPodsInNamespaceAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.BridgeToKubernetes.Common.Kubernetes.KubernetesClient.<>c__DisplayClass75_0`1.<<ClientInvokeWrapperAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.BridgeToKubernetes.Common.Utilities.WebUtilities.RetryUntilTimeAsync(Func`2 action, TimeSpan maxWaitTime, CancellationToken cancellationToken)\n   at Microsoft.BridgeToKubernetes.Common.Kubernetes.KubernetesClient.ClientInvokeWrapperAsync[T](Func`1 handler, String operation, CancellationToken cancellationToken)\n   at Microsoft.BridgeToKubernetes.Common.Kubernetes.KubernetesClient.ListPodsInNamespaceAsync(String namespaceName, IEnumerable`1 labels, CancellationToken cancellationToken)\n   at Microsoft.BridgeToKubernetes.Library.Client.ManagementClients.KubernetesManagementClient.<>c__DisplayClass6_0.<<RefreshCredentialsAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ManagementClientExceptionStrategy.<>c__DisplayClass3_0.<<RunWithHandlingAsync>b__0>d.MoveNext()\n--- End of stack trace from previous location where exception was thrown ---\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ManagementClientExceptionStrategy.RunWithHandlingAsync[T](Func`1 func, FailureConfig failureConfig)",
    "Message": "The SSL connection could not be established, see inner exception.",
    "Data": {},
    "InnerException": {
        "ClassName": "System.Security.Authentication.AuthenticationException",
        "Message": "Authentication failed, see inner exception.",
        "Data": null,
        "InnerException": {
            "StackTrace": null,
            "Message": "Internal error",
            "Data": {},
            "InnerException": null,
            "HelpLink": null,
            "Source": null,
            "HResult": -9810
        },
        "HelpURL": null,
        "StackTraceString": "   at System.Net.Security.SslStream.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)\n   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)\n   at System.Net.Security.SslStream.PartialFrameCallback(AsyncProtocolRequest asyncRequest)\n--- End of stack trace from previous location where exception was thrown ---\n   at System.Net.Security.SslStream.ThrowIfExceptional()\n   at System.Net.Security.SslStream.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)\n   at System.Net.Security.SslStream.EndProcessAuthentication(IAsyncResult result)\n   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)\n   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__65_1(IAsyncResult iar)\n   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)\n--- End of stack trace from previous location where exception was thrown ---\n   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)",
        "RemoteStackTraceString": null,
        "RemoteStackIndex": 0,
        "ExceptionMethod": null,
        "HResult": -2146233087,
        "Source": "System.Private.CoreLib",
        "WatsonBuckets": null
    },
    "HelpLink": null,
    "Source": "System.Net.Http",
    "HResult": -2146233087
}
```

My dotnet runtime: 
```
% dotnet --list-runtimes
Microsoft.AspNetCore.App 3.1.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 3.1.17 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

OS: 
```
MacOS Big Sur 11.5.1 (20G80)
```

I've also tried the guidance mentioned here but to no avail: https://github.com/dotnet/runtime/issues/27132#issuecomment-419232421
I didn't see any errors in the output when running `sudo dotnet dev-certs https --clean` or `dotnet dev-certs https -t`.
Manually deleting the localhost certificates from the Keychain didn't work either.

Any idea on why this reproduces after such a seemingly innocuous change? 
Thank you in advance!

**Comment by wfurt** [comment]
What is the inner Exception  @amsoedal? This is somewhat hard to decode from the post. 
Now, is there chance that there is certificate where the name or host changes based on the project name? 

It seems like your best chance would be small repro with just HttpClient or SslStream. You can also try to get packet captures with Wireshark and compare the good and bad handshakes. (or post the traces)

**Comment by amsoedal** [comment]
Hi @wfurt, thanks so much for the reply. The inner exception is not much better than what I posted in the logs above: 

Interop+AppleCrypto+SslException: {"StackTrace":null,"Message":"Internal error","Data":{},"InnerException":null,"HelpLink":null,"Source":null,"HResult":-9810}

My colleague and I have been digging through our code + pipelines but we haven't found anything dependent on the project name that hasn't been updated (nor have I found anything in the certs on my machine, although I'm guessing that's where the problem lies -- in some MacOS configuration). 

To get a small repro, I replaced our application logic with the logic from this sample app (https://github.com/dotnet/samples/tree/main/csharp/getting-started/console-webapiclient) and rebuilt the bits from the pipeline. Everything worked. (Does this mean the issue is related to the KubernetesClient?) 

Basically, it only reproduces on a few machines, after building from the pipeline instead of locally, and when trying to talk to a K8s cluster via the KubernetesClient. I'll follow up with you on Teams re: the Wireshark traces

**Comment by wfurt** [comment]
That error is `errSSLInternal` so it is not very helpful. I look at the captures and there seems to be issue with client certificate. In the both  cases, server would respond and send request for client certificate.  SNI and other parts look good to me. In the successful case client would present own certificate but in the failing case it just closes the connection. 

I would check if the client cert can be loaded and has accessible private key. macOS is also sensitive to key partition and signing so for example same app signed and unsigned may be able to access different keys. 

You can also look at system logs. Most interesting from `securityd` - the primary guardian of secrets. 
```
ps axu | grep securityd
root               521   0.0  0.0  4613716    196   ??  Ss   30Jun21   0:00.14 /usr/libexec/securityd_service
root               181   0.0  0.0  6437256   6040   ??  Ss   31Dec69  12:56.66 /usr/sbin/securityd -i
furt             81172   0.0  0.0  4427004    852 s036  S+   11:38AM   0:00.00 grep securityd
Shining:libraries furt$ log stream --process 181
```

**Comment by ManickaP** [comment]
Triage: once we know more, we should figure out if it's ours fault or not for 7.0.

**Comment by amsoedal** [comment]
@wfurt thanks so much. Just got around to trying this today, and the logs were revealing:

```
2021-08-06 12:33:46.811430+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:clientid] code requirement check failed (-67063), client is not Apple-signed
2021-08-06 12:33:46.811497+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:kcacl] Keychain query for process 10880 (UID 503)
2021-08-06 12:33:46.811535+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:security_exception] MacOS error: -67063
2021-08-06 12:33:46.811596+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:kcacl] client is invalid, suppressing prompt
2021-08-06 12:33:46.811672+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:kcacl] suppressing keychain prompt for invalidly signed client /Users/ansoedal/Library/Application Support/Code/User/globalStorage/mindaro.mindaro/file-downloader-downloads/binaries/bridge(10880)
2021-08-06 12:33:46.811749+0100 0x27f      Default     0x0                  120    0    securityd: (Security) [com.apple.securityd:security_exception] MacOS error: -25337
2021-08-06 12:33:46.811864+0100 0x27f      Default     0x0                  120    0    securityd: [com.apple.securityd:security_exception] CSSM Exception: 32 unknown error 32=20
```

We've known we had signing issues, but I didn't think this change would cause them to surface, since nothing changed anything about our signing or our pipeline. Thanks so much for your help, closing this since it's not a dotnet issue but rather a signing issue on our end.
